### PR TITLE
🎨 Introduce product in api server

### PR DESCRIPTION
**Pull request by bisgaard-itis**
## What do these changes do?
- Introduce products to the API server
- Ensure product is sent in header to webserver. N.B. the current PR only handles the webserver. Other backend services will be handled in a separate PR

## Related issue/s
- https://github.com/ITISFoundation/osparc-simcore/issues/4914

### CODE REVIEW

**Review by pcrespov — COMMENTED** [review]
I left some suggestions

File: services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py [hunk 1]
```diff
@@ -25,27 +31,35 @@ def _create_exception() -> HTTPException:
...
+) -> Identity:
+    user_and_product: UserAndProductTuple | None = await apikeys_repo.get_user(
         api_key=credentials.username, api_secret=credentials.password
     )
-    if not user_id:
+    if user_and_product is None:
         exc = _create_exception()
         raise exc
-    return user_id
+    email = await users_repo.get_email_from_user_id(user_and_product[0])
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py:47]: Good point. I will fix

File: services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py [hunk 2]
```diff
@@ -25,27 +31,35 @@ def _create_exception() -> HTTPException:
...
+    if not email:
+        exc = _create_exception()
+        raise exc
+    return Identity(
+        user_id=user_and_product[0], product_name=user_and_product[1], email=email
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py:52]: Good point. I will adapt

File: services/api-server/src/simcore_service_api_server/api/dependencies/webserver.py [hunk 3]
```diff
@@ -61,12 +61,16 @@ def get_session_cookie(
...
     session_cookies: Annotated[dict, Depends(get_session_cookie)],
+    identity: Annotated[Identity, Depends(get_current_identity)],
 ) -> AuthSession:
     """
     Lifetime of AuthSession wrapper is one request because it needs different session cookies
     Lifetime of embedded client is attached to the app lifetime
     """
-    session = AuthSession.create(app, session_cookies)
+    product_header: dict[str, str] = {
+        "X-Simcore-Products-Name": f"{identity.product_name}"
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply services/api-server/src/simcore_service_api_server/api/dependencies/webserver.py:71]: Yep, I moved it into `servicelib.rest_constants` as discussed online

File: services/api-server/tests/unit/_with_db/conftest.py [hunk 4]
```diff
@@ -176,48 +177,97 @@ async def connection(app: FastAPI) -> AsyncIterator[SAConnection]:
...
+                if entry is None:
+                    break
+            uid = await connection.scalar(
+                users.insert().values(user).returning(users.c.id)
+            )
+            assert uid
+            _generate_user_ids.generated_ids.append(uid)
+            yield uid
+
+    _generate_user_ids.generated_ids = []
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply services/api-server/tests/unit/_with_db/conftest.py:200]: Good point. I will fix

File: services/api-server/tests/unit/_with_db/test_product.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+async def test_webserver_product(
+    client: httpx.AsyncClient,
+    mocked_webserver_service_api_base: respx.MockRouter,
+    fake_api_keys: Callable[[PositiveInt], AsyncGenerator[ApiKeyInDB, None]],
+    faker: Faker,
+) -> None:
+    assert client
+
+    keys: dict[int, ApiKeyInDB] = {}
```
[the remark this answers is not in the document]
Comment (bisgaard-itis) [reply services/api-server/tests/unit/_with_db/test_product.py:26]: OK, I will try to get the products set up for the the other backends for now and setup some more testing, then I hope I can clean this up at some point. I already spent quite some time trying to adapt the fixtures to have multiple users/identities in the DB.

**Review by pcrespov — APPROVED** [review]
:tada:

File: packages/service-library/src/servicelib/rest_constants.py [hunk 6]
```diff
@@ -16,3 +16,6 @@ class PydanticExportParametersDict(TypedDict):
     exclude_defaults=False,
     exclude_none=False,
 )
+
+# Headers keys
+X_PRODUCT_NAME_HEADER: Final[str] = "X-Simcore-Products-Name"
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/rest_constants.py:21]: minor: isn't there some singular vs plural issue here?

File: services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py [hunk 7]
```diff
@@ -2,16 +2,22 @@
...
 from .database import get_repository
 
 # SEE https://swagger.io/docs/specification/authentication/basic-authentication/
 basic_scheme = HTTPBasic()
 
 
+class Identity(BaseModel):
+    user_id: PositiveInt
+    product_name: str
+    email: str
```
Comment (sanderegg) [inline services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py:18]: pydantic has a validator for email, something like EmailStr and I see we also have LowerCaseEmailStr... maybe worth checking?

**Review by sanderegg — APPROVED** [review]
nice! just minor comments, feel free to check

File: packages/service-library/src/servicelib/rest_constants.py (hunk 6 above)
Comment (bisgaard-itis) [reply packages/service-library/src/servicelib/rest_constants.py:21]: I agree. But I just moved this one from the webserver to here, so I don't want to mess with it. That would surely cause issues

File: services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py (hunk 7 above)
Comment (bisgaard-itis) [reply services/api-server/src/simcore_service_api_server/api/dependencies/authentication.py:18]: Good point. I will make that change in a second follow-up PR about the products in the api server

**Pull request merged**